pkg/storage/tsdb/config.go

```diff
@@ -176,6 +179,7 @@ func (cfg *TSDBConfig) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.CloseIdleTSDBTimeout, "blocks-storage.tsdb.close-idle-tsdb-timeout", 0, "If TSDB has not received any data for this duration, and all blocks from TSDB have been shipped, TSDB is closed and deleted from local disk. If set to positive value, this value should be equal or higher than -querier.query-ingesters-within flag to make sure that TSDB is not closed prematurely, which could cause partial query results. 0 or negative value disables closing of idle TSDB.")
 	f.IntVar(&cfg.MaxExemplars, "blocks-storage.tsdb.max-exemplars", 0, "Enables support for exemplars in TSDB and sets the maximum number that will be stored. 0 or less means disabled.")
 	f.IntVar(&cfg.HeadChunksWriteQueueSize, "blocks-storage.tsdb.head-chunks-write-queue-size", chunks.DefaultWriteQueueSize, "The size of the in-memory queue used before flushing chunks to the disk.")
+	f.BoolVar(&cfg.MemorySnapshotOnShutdown, "blocks-storage.tsdb.memory-snapshot-on-shutdown", false, "True to enable snapshotting of in-memory TSDB data on disk when shutting down.")
```

Comment:
I think it is better to make this a per tenant flag rather than a global flag. Can we add it to https://github.com/cortexproject/cortex/blob/master/pkg/util/validation/limits.go?
It is useful for testing purpose if we want to enable it for some tenants and not all tenants first. Once we think this is stable, it is also easy to enable it globally for all tenants by changing the default value.

Reply:
The purpose of snapshots is to speed up node recovery. From the perspective of cluster users, all tenants should be equal. For example, when there are hundreds of tenants in the cluster, you need to fill in the ID of each tenant if you want it to take effect, otherwise (a small number of tenants take effect) the acceleration effect will be weakened.

Reply:
As I said, if users want to enable it for all tenants then they can just set the default value to true to enable all and this is the same as what you are doing right now.
Making it a per tenant config makes it more flexible for testing and it is much safer as an experimental feature.